### ISSUE TITLE
disable_os_cache on Windows does not disable file system cache?

### ISSUE THREAD

**Issue by hg** [arvidn/libtorrent#4011]
libtorrent version (or branch): 1.1.13.0/any
platform/architecture: amd64/any
compiler and compiler version: MSVC 15.3 (?)/any

Hi,

I've been using qBittorrent and noticed that the "disable OS cache" option seems to be doing nothing: the system still eats up all available RAM for file system cache (freshly booted, with SuperFetch/prefetcher disabled and no other application running).

I dug a little into libtorrent source code and to disable file system cache on Windows [it seems to be using](https://github.com/arvidn/libtorrent/blob/d33b0506a0f4cd61860fd54864267c82d3e171e3/src/file.cpp#L531) the `FILE_FLAG_WRITE_THROUGH` flag. I think it's not enough and `FILE_FLAG_NO_BUFFERING` should be added to it: [doc 1](https://support.microsoft.com/en-us/help/99794/info-file-flag-write-through-and-file-flag-no-buffering), [doc 2](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilew).

Here's the relevant excerpt from the first link:
>The FILE_FLAG_WRITE_THROUGH flag for CreateFile() causes any writes made to that handle to be written directly to the file without being buffered. The data is cached (stored in the disk cache); however, it is still written directly to the file.
>The FILE_FLAG_NO_BUFFERING takes this concept one step further and eliminates all read-ahead file buffering and disk caching as well, so that all reads are guaranteed to come from the file and not from any system buffer or disk cache. 

And the second:
>If FILE_FLAG_WRITE_THROUGH is used but FILE_FLAG_NO_BUFFERING is not also specified, so that system caching is in effect, then the data is written to the system cache but is flushed to disk without delay.
>If FILE_FLAG_WRITE_THROUGH and FILE_FLAG_NO_BUFFERING are both specified, so that system caching is not in effect, then the data is immediately flushed to disk without going through the Windows system cache. The operating system also requests a write-through of the hard disk's local hardware cache to persistent media.

I would have tried to patch the code myself and submit a pull request, but this one got me worried:
>When using FILE_FLAG_NO_BUFFERING, disk reads and writes must be done on sector boundaries, and buffer addresses must be aligned on disk sector boundaries in memory.

I am afraid I am not qualified enough to make sure the project is not broken after such a change. So I wrote [a small program](https://gist.github.com/hg/aca977c2029875b04d6c9d055be354e2) to test both flags. It creates a temporary file, writes zeroes to it, and reads them back, once for each of the above flags. The FS cache was flushed manually between each iteration (using `RamMap`, `Menu` -> `Empty` -> `Empty Standby List`; it could be done with code, but it adds two times the boilerplate due to privilege handling).

Here's the results (`Cached` value from `Task Manager` -> `Performance` -> `Memory`):

- Writing with `FILE_FLAG_WRITE_THROUGH`, before/after: 130 MB -> 4.2 GB
- Reading with `FILE_FLAG_WRITE_THROUGH`, before/after: 72 MB -> 4.2 GB
- Writing with `FILE_FLAG_NO_BUFFERING`, before/after: 71 MB -> 93 MB
- Reading with `FILE_FLAG_NO_BUFFERING`, before/after: 69 MB -> 70 MB

So they seem to behave the way quoted above.

Would any of the core developers kindly take a look at this?

Thanks!

**Comment by arvidn** [comment]
The option to disable OS cache in libtorrent was deprecated and has been removed in newer versions.
making libtorrent clear the filesystem cache is problematic. primarily because it requires privileges, but also because a library in one process shouldn't really perform system wide operations like that.

This topic has been discussed quite a bit in the past. It all comes down to:

what is the problem with windows using all your RAM for file system cache?

**Comment by arvidn** [comment]
btw, libtorrent used to use `FILE_FLAG_NO_BUFFERING` a long time ago, but the complexity and performance degradation was unsustainable.

for reference:
https://github.com/arvidn/libtorrent/issues/3614
https://github.com/arvidn/libtorrent/issues/3687
https://github.com/qbittorrent/qBittorrent/issues/8328
https://github.com/qbittorrent/qBittorrent/issues/5804

**Comment by hg** [comment]
I see.

>what is the problem with windows using all your RAM for file system cache?

In my experience, torrent clients tend to pollute FS cache with torrent data forcing the more useful (but rarely used) application data out. If I switch to a dictionary or to an IDE which hasn't been used for an hour, the response time and sluggishness is pretty bad if there was a torrent client running.

On Linux it's pretty easy to mitigate via per-process memory limits which can be configured with cgroups (just stick the process into a separate cgroup and set the `memory.max_usage_in_bytes` knob to whatever your heart desires). I set mine to 2 GBs and it had zero impact on upload/download speeds, but when switching to background applications the difference is night and day.

Looks like there's no way to do the same on Windows, there's only a global limit for the whole system.

Thanks for your help!

**Comment by hg** [comment]
TLDR: set qBittorrent's memory priority to low.

---

Sorry to keep badgering your issue tracker with things which are hardly relevant to libtorrent, but I must share the solution to this problem for future reference.

[This article](https://habrahabr.ru/blogs/windows7/107637/) (in Russian), written by an experienced driver developer, suggests lowering the I/O and memory priorities of the torrent process to avoid pushing more useful data out of the FS cache. The author also links to [this document](http://download.microsoft.com/download/a/f/7/af7777e5-7dcd-4800-8a0a-b18336565f5b/Priorityio.doc) (in English).

This easiest way to do it is to import the snippet below into the registry (it also lowers qBittorrent's CPU priority for good measure), then either reboot the machine, or restart qBittorrent and flush the standby lists with RamMap. This only needs to be done once.

```
Windows Registry Editor Version 5.00

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\Image File Execution Options\qbittorrent.exe]

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\Image File Execution Options\qbittorrent.exe\PerfOptions]
"CpuPriorityClass"=dword:00000001
"IoPriority"=dword:00000000
"PagePriority"=dword:00000001
```

I've been using this for about a day and it works extremely well in practice. Here's how I checked it:

1. rebooted to get a clean system;
1. opened up a bunch of video files (not used by any torrents), forcing them into RAM by playing them back and forth; then started a heavy application, used it for a while, and closed it;
1. went into `Process Explorer -> Ctrl+I -> Memory` and recorded the standby list sizes (~50 MB for priority 1, ~3 GB priority 5).
1. started qB, added a few torrents with total size 4 times the amount of RAM on my machine in addition to what's already been there;
1. left it working for two hours with qB the only application running;
1. stopped qB 

The priority 5 standby list is still ~3 GB in size, while priority 1 took all available RAM. I then went into RamMap and checked the cache details -- my video and application files from step 2 are still there. The system stayed very responsive and all sluggishness is gone. Before applying this fix, the useful data would have been forced out in 5-10 minutes, which I have seen so many times before.

If I start any applications or force any files into cache, the priority 5 standby list starts growing and eating memory from priority 1 (i.e. the torrent client).

Could lowering the memory priority be integrated into libtorrent, or is this the wrong place to ask and this could/should be moved to the client itself? Although this works very nicely for me, further testing needs to be done to determine how the system behaves when higher priority lists leave no RAM for the lower ones.

See [SetProcessInformation](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-setprocessinformation) for more info.

**Comment by arvidn** [comment]
Libtorrent has an option to set disk IO. https://github.com/arvidn/libtorrent/blob/RC_1_2/src/file_pool.cpp#L54

I would be reluctant to make calls that affect the whole process, and the documentation on msdn is a bit ambiguous on whether the call is process wide or per thread. But it seems like a good idea for clients to do. Especially given disk io will use memory mapped files in the next major release

**Comment by Chocobo1** [comment]
>and the documentation on msdn is a bit ambiguous on whether the call is process wide or per thread. 

There are [SetProcessInformation](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-setprocessinformation) and [SetThreadInformation](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-setthreadinformation). Now I wonder if there is a way to use `SetThreadInformation` on libtorrent threads? Also for optimal results, I suspect not every libtorrent threads should be tweaked, only the ones related to the storage.
